### PR TITLE
coredump: log minimal metadata early

### PR DESCRIPTION
**Pull request by hb-mvrd**
If a process crashes during shutdown, the writing of the coredump may be interrupted by the shutdown and only the PID of the crashed process makes it into the journal.

Logging the executable path and signal before writing the coredump ensures that this information is available.

### CODE REVIEW

File: src/coredump/coredump.c [hunk 1]
```diff
@@ -1587,6 +1587,14 @@ static int process_kernel(int argc, char* argv[]) {
                 /* OK, now we know it's not the journal, hence we can make use of it now. */
                 log_set_target_and_open(LOG_TARGET_JOURNAL_OR_KMSG);
 
+        /* Log minimal metadata now, so it is not lost if the system is about to shut down. */
+        log_info("PID %s (%s) crashed with signal %s/%s",
+                 context.meta[META_ARGV_PID],
```
Comment (poettering) [inline src/coredump/coredump.c]: we usually indent continuation lines to twice the usual indentation, hence 16 spaces, please

File: src/coredump/coredump.c [hunk 2]
```diff
@@ -1587,6 +1587,14 @@ static int process_kernel(int argc, char* argv[]) {
...
+                 context.meta[META_EXE] ? context.meta[META_EXE] : "unknown",
```
Comment (poettering) [inline src/coredump/coredump.c]: strna(context.meta[META_EXE])

File: src/coredump/coredump.c [hunk 3]
```diff
@@ -1587,6 +1587,14 @@ static int process_kernel(int argc, char* argv[]) {
...
+                 context.meta[META_ARGV_SIGNAL],
+                 (safe_atoi(context.meta[META_ARGV_SIGNAL], &signo) >= 0 && SIGNAL_VALID(signo))
+                 ? signal_to_string(signo) : "UNKNOWN");
```
Comment (poettering) [inline src/coredump/coredump.c]: strna(signal_to_string(signo))

File: src/coredump/coredump.c [hunk 4]
```diff
@@ -1587,6 +1587,14 @@ static int process_kernel(int argc, char* argv[]) {
                 /* OK, now we know it's not the journal, hence we can make use of it now. */
                 log_set_target_and_open(LOG_TARGET_JOURNAL_OR_KMSG);
 
+        /* Log minimal metadata now, so it is not lost if the system is about to shut down. */
+        log_info("PID %s (%s) crashed with signal %s/%s",
+                 context.meta[META_ARGV_PID],
+                 context.meta[META_EXE] ? context.meta[META_EXE] : "unknown",
+                 context.meta[META_ARGV_SIGNAL],
+                 (safe_atoi(context.meta[META_ARGV_SIGNAL], &signo) >= 0 && SIGNAL_VALID(signo))
```
Comment (poettering) [inline src/coredump/coredump.c]: signal_to_string() is happily handles signal numbers that are out of range, hence you can drop the extra SIGNAL_VALID() check here

File: src/coredump/coredump.c [hunk 5]
```diff
@@ -1587,6 +1587,14 @@ static int process_kernel(int argc, char* argv[]) {
...
+                 ? signal_to_string(signo) : "UNKNOWN");
+
```
Comment (poettering) [inline src/coredump/coredump.c:1595]: so i think adding a log message like this is OK, but I think we should reword this to say that we are beginning to process the coredump now. That way the duplicate log message is less confusing: we'll log one message when we start processing it, and another when we are done, and that should become clear from the text

File: src/coredump/coredump.c [hunk 6]
```diff
@@ -1587,6 +1587,14 @@ static int process_kernel(int argc, char* argv[]) {
                 /* OK, now we know it's not the journal, hence we can make use of it now. */
                 log_set_target_and_open(LOG_TARGET_JOURNAL_OR_KMSG);
 
+        /* Log minimal metadata now, so it is not lost if the system is about to shut down. */
+        safe_atoi(context.meta[META_ARGV_SIGNAL], &signo);
```
Comment (poettering) [inline src/coredump/coredump.c]: this line is redundant, as you have the same expression below

File: src/coredump/coredump.c [hunk 7]
```diff
@@ -1587,6 +1587,14 @@ static int process_kernel(int argc, char* argv[]) {
...
+        log_info("PID %s (%s) crashed with signal %s/%s, processing coredump.",
+                        context.meta[META_ARGV_PID], strna(context.meta[META_EXE]),
+                        context.meta[META_ARGV_SIGNAL],
+                        safe_atoi(context.meta[META_ARGV_SIGNAL], &signo) >= 0
+                        ? strna(signal_to_string(signo)) : "n/a");
```
Comment (poettering) [inline src/coredump/coredump.c]: strna() already generates the `"n/a"` string, hence change to:

```c
strna(safe_atoi(…, &signo) >= 0 ? signal_to_string(signo) : NULL)
```

File: src/coredump/coredump.c [hunk 8]
```diff
@@ -1587,6 +1587,14 @@ static int process_kernel(int argc, char* argv[]) {
                 /* OK, now we know it's not the journal, hence we can make use of it now. */
                 log_set_target_and_open(LOG_TARGET_JOURNAL_OR_KMSG);
 
+        /* Log minimal metadata now, so it is not lost if the system is about to shut down. */
+        safe_atoi(context.meta[META_ARGV_SIGNAL], &signo);
+        log_info("PID %s (%s) crashed with signal %s/%s, processing coredump.",
```
Comment (poettering) [inline src/coredump/coredump.c]: So this is still a bit too similar (and also too different) to me to the message we write when processing is done, which is:

```
        core_message = strjoin(
                        "Process ", context->meta[META_ARGV_PID],
                        " (", context->meta[META_COMM],
                        ") of user ", context->meta[META_ARGV_UID],
                        written ? " dumped core." : " terminated abnormally without generating a coredump.");
```

Also, I don#t like the word "crashed", because it's not clear what that means.

Please make this somewhat symmetric and use similar wording.

maybe something like this:

start: `"Process xyz (abc) of user def terminated abnormally, processing…"`
done: `"Process xyz (abc) of user def dumped core."`

**Comment by hb-mvrd** [comment]
I'd like to keep the signal number and name in the log. Updated message:
start: `Process 515 (nginx) of user 1002 terminated with signal 11/SEGV, processing...`
done: `Process 515 (nginx) of user 1002 dumped core.`

File: src/coredump/coredump.c [hunk 9]
```diff
@@ -1587,6 +1587,12 @@ static int process_kernel(int argc, char* argv[]) {
                 /* OK, now we know it's not the journal, hence we can make use of it now. */
                 log_set_target_and_open(LOG_TARGET_JOURNAL_OR_KMSG);
 
+        /* Log minimal metadata now, so it is not lost if the system is about to shut down. */
+        log_info("Process %s (%s) of user %s terminated with signal %s/%s, processing...",
```
Comment (poettering) [inline src/coredump/coredump.c]: i think we need to say *terminated* *abnormally*. otherwise it's not obvious why we are doing anything

**Comment by hb-mvrd** [comment]
Updated message:
start: `Process 515 (nginx) of user 1002 terminated abnormally with signal 11/SEGV, processing...`
done: `Process 515 (nginx) of user 1002 dumped core.`
